pylot/prediction/r2p2_predictor_operator.py

```diff
@@ -0,0 +1,172 @@
+        ego_vehicle_trajectory = np.stack(
+            [[point.location.x, point.location.y] for
+              point in ego_vehicle_trajectory])
+
+        # If we haven't seen enough past locations yet, pad the start
+        # of the trajectory with copies of the earliest location.
+        num_past_locations = ego_vehicle_trajectory.shape[0]
+        if num_past_locations < self._flags.prediction_num_past_steps:
+            initial_copies = np.repeat([np.array(ego_vehicle_trajectory[0])],
+                                       self._flags.prediction_num_past_steps - num_past_locations,
+                                       axis=0)
+            ego_vehicle_trajectory = np.vstack(
+                (initial_copies, ego_vehicle_trajectory))
+        ego_vehicle_trajectory = np.expand_dims(ego_vehicle_trajectory, axis=0)
+
+        z = torch.tensor(np.random.normal(
+            size=(1, self._flags.prediction_num_future_steps, 2))).to(
+            torch.float32).to(self._device)
+        ego_vehicle_trajectory = torch.tensor(ego_vehicle_trajectory).to(torch.float32).to(self._device)
+        binned_lidar = torch.tensor(binned_lidar).to(torch.float32).to(self._device)
+
+        prediction_array, _ = self._r2p2_model.forward(z,
+                                  ego_vehicle_trajectory,
+                                  binned_lidar)
+        prediction_array = prediction_array.cpu().detach().numpy()[0]
+
+        # Convert prediction array to a list of Location objects.
+        predictions = []
+        for t in range(self._flags.prediction_num_future_steps):
+            predictions.append(
+                Transform(location=Location(x=prediction_array[t][0],
+                                            y=prediction_array[t][1]),
+                          rotation=Rotation()))
+
+        obstacle_predictions_list = []
+        obstacle_predictions_list.append(
+            ObstaclePrediction('vehicle',
+                               self.vehicle_id,
+                               self._vehicle_transform,
+                               ego_vehicle_bbox,
+                               1.0, # Probability; currently a filler value because we are taking just one sample from distribution
+                               predictions))
+        prediction_stream.send(
+            PredictionMessage(timestamp, obstacle_predictions_list))
+
+    def get_occupancy_grid(self, point_cloud):
+        """Get occupancy grids at two different heights."""
+
+        z_threshold = -3.4 # Threshold used in the PRECOG (https://arxiv.org/pdf/1905.01296.pdf) dataset
+
+        above_mask = point_cloud[:, 2] > z_threshold
+        #print ("Points above threshold:", sum(above_mask))
+        #print ("Points below threshold:", sum(1 - above_mask))
+
+        def get_occupancy_from_masked_lidar(mask):
+            masked_lidar = point_cloud[mask]
+            meters_max = 50
```

Comment:
Shouldn't we get this from the lidar setup? https://github.com/erdos-project/pylot/blob/master/pylot/drivers/sensor_setup.py#L156